### PR TITLE
Use Json Schema for values validation

### PR DESCRIPTION
**Pull request by abrahamko**
### What does this PR do?
- _What's changed? Why were these changes made?_

Validate values using Json Schema to improve UX, instead of inline validation.
Benefits:
 1. Validation resides next to the values.yaml instead of inline in the templates
 2. Error handling done by Json Schema infra
 3. Use built-in validation tool
 4. Can validate types of input, range, regex and more
 5. All validations are concentrated in one place

NOTE:
`required` properties were used only for the ones that were already `required`.
Besides that, only type validation was used, with minimal `1` for expiration properties.

- _How should the reviewer approach this PR, especially if manual tests are required?_
1. Check that the existing validations are maintained
2. Execute `helm template` to make sure the existing values pass 

- _Are there relevant screenshots you can add to the PR description?_ NO

### What ticket does this PR close?
Connected to #69 

#### Change log
- [x] The CHANGELOG has been updated, or
- [ ] This PR does not include user-facing changes and doesn't require a CHANGELOG update

#### Test coverage
- [ ] This PR includes new unit and integration tests to go with the code changes, or
- [x] The changes in this PR do not require tests

#### Documentation
- [ ] Docs (e.g. `README`s) were updated in this PR, and/or there is a follow-on issue to update docs, or
- [x] This PR does not require updating any documentation

### CODE REVIEW

File: conjur-oss/values.yaml [hunk 1]
```diff
@@ -120,10 +120,6 @@ service:
     port: 443
     type: NodePort
 
-serviceAccount:
-  create: true
-  name:
-
```
Comment (abrahamko) [inline conjur-oss/values.yaml:126]: Duplicated. Same lines with comment are below.

File: conjur-oss/values.schema.json [hunk 2]
```diff
@@ -0,0 +1,234 @@
+{
+  "$schema": "http://json-schema.org/draft-07/schema",
+  "required": [
+    "database",
+    "dataKey",
+    "ssl"
```
Comment (sgnn7) [inline conjur-oss/values.schema.json:6]: I think database and ssl are conditionally required and as such shouldn't be required in the schema

Comment (diverdane) [reply conjur-oss/values.schema.json:6]: Yes, the `database` and `ssl` sections are optional. For example, if you don't supply any `database.*` values for `helm install`, then the Helm chart will create a database for you with autogenerated SSL certificates & an autogenerated password. Similarly, if you don't supply any `ssl.*` values, the Helm chart will autogenerate SSL certificates and a CA cert chain for Conjur access.

However, there are some "conditionally required" values/fields within the `database` and `ssl` sections. For example, if someone supplies `database.ssl.cert`, then we want to make sure that they also supply `database.ssl.key`. I'll add more comments below for these "conditionally required" fields.

File: conjur-oss/values.schema.json [hunk 3]
```diff
@@ -0,0 +1,234 @@
...
+    "database": {
+      "required": [
+        "ssl"
+      ],
+      "properties": {
+        "password": {
+          "type": "string"
+        },
+        "ssl": {
+          "required": [
```
Comment (diverdane) [inline conjur-oss/values.schema.json:30]: The `ssl.key` value should only be required if `ssl.cert` is set. I think you can use the `dependencies` keyword (see https://stackoverflow.com/questions/38717933/jsonschema-attribute-conditionally-required); e.g. add something like this (not sure if I have the syntax correct):
```
"dependencies": {
  "cert": { "required": ["key"] }
}
```
at line 45.

File: conjur-oss/values.schema.json [hunk 4]
```diff
@@ -0,0 +1,234 @@
...
+          "type": "boolean"
+        },
+        "name": {
+          "type": "string"
+        }
+      }
+    },
+    "ssl": {
+      "required": [
+        "caKey",
```
Comment (diverdane) [inline conjur-oss/values.schema.json:185]: The `caKey`, `cert`, and `key` fields here should be conditionally required.
Thinking about this a bit, I believe the logic should be that `caKey`, `caCert`, and `key` are required if `cert` is set.

Again, I think you can use the `dependencies` keyword (see https://stackoverflow.com/questions/38717933/jsonschema-attribute-conditionally-required); e.g. add something like this (not sure if I have the syntax correct):
```
"dependencies": {
  "cert": { "required": ["key", "caCert", "caKey"] }
}
```
at line 205.

File: conjur-oss/values.schema.json [hunk 5]
```diff
@@ -0,0 +1,234 @@
...
+    },
+    "authenticators": {
+      "type": "string"
+    },
+    "conjurLabels": {
+      "type": "object"
+    },
+    "database": {
+      "required": [
+        "ssl"
```
Comment (diverdane) [inline conjur-oss/values.schema.json:23]: The "ssl" field isn't required. If you don't provide this, the chart will create an SSL cert for you.

File: conjur-oss/values.schema.json [hunk 6]
```diff
@@ -0,0 +1,234 @@
...
+        "persistentVolume": {
+          "properties": {
+            "create": {
+              "type": "boolean"
+            },
+            "size": {
+              "type": "string"
+            },
+            "storageClass": {
+              "type": "string"
```
Comment (diverdane) [inline conjur-oss/values.schema.json:120]: For some reason, adding this "string" type for "storageClass" causes errors when "storageClass" isn't provided. I think the fix is to comment out the "storageClass line in the values.yaml file. That is, replace:
```
    storageClass:
```
with:
```
    # Set storageClass to use a non-default storage class for your platform
    # storageClass:
```

File: conjur-oss/values.schema.json [hunk 7]
```diff
@@ -0,0 +1,234 @@
...
+        }
+      }
+    },
+    "serviceAccount": {
+      "properties": {
+        "create": {
+          "type": "boolean"
+        },
+        "name": {
+          "type": "string"
```
Comment (diverdane) [inline conjur-oss/values.schema.json:179]: For some reason, declaring "name" field as type "string" causes an error when "string" isn't supplied. I think the fix is to comment out the "name" field in values.yaml, e.g. replace this:
```
  name:
```
with:
```
  # name:
```

**Comment by diverdane** [comment]
Hi @abrahamko ,
Have you had a chance to look at my review comments?
Please let me know if I can help on this, this is a good change to get in.
Cheers,
Dane

**Comment by abrahamko** [comment]
Hi @diverdane ,
Thanks for the detailed review.
I handled your comments as you suggested, and verified they behave as you would expect.
Still, since there are defaults in the values.yaml that does not comply with the schema, I had to comment them out too. (the `cert`, `key` etc.)
The Helm Linter still fails because there's a bug in the linter that prevents it from using the `--set`, so it only sees the default (which is empty).
This was fixed in latest helm release `3.3.0`, which was released 5 days ago ;)
If you can upgrade the helm used to do the lint verification, it should be removed.

File: CHANGELOG.md [hunk 8]
```diff
@@ -4,6 +4,10 @@ All notable changes to this project will be documented in this file.
 The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
 and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).
 
+## [Unreleased]
+### Changed
+- Use Json Schema for values validation instead of template `required` function
```
Comment (izgeri) [inline CHANGELOG.md:9]: Is there an end-user benefit or impact from this change? If not, I don't think we need to include it in the changelog. Also, all change log entries must have an issue link now. See our [changelog guidelines](https://github.com/cyberark/community/blob/master/Conjur/CONTRIBUTING.md#changelog-guidelines) for detailed info.

**Comment by diverdane** [comment]
Hi @abrahamko, 

That Helm Linter error is a Linter bug that was resolved in Helm 3.3.0:
https://github.com/helm/helm/issues/7273

I don't know if the Github actions for Helm linter will run with Helm 3.3.0, but can you try this change:

```
diff --git a/.github/workflows/ci.yml b/.github/workflows/ci.yml
index 955b5a2..164920a 100644
--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -15,7 +15,8 @@ jobs:
       - name: Install Helm
         uses: azure/setup-helm@v1
         with:
-          version: v3.2.1
+          # Use Helm v3.3.0 to include helm/helm PR #7984
+          version: v3.3.0
 
       - name: Run linter
         run: helm lint --set dataKey=BLOB ./conjur-oss
diff --git a/conjur-oss/values.yaml b/conjur-oss/values.yaml
index fb19141..966313b 100644
--- a/conjur-oss/values.yaml
+++ b/conjur-oss/values.yaml
@@ -47,7 +47,8 @@ database:
 
 # Conjur data key, 32 byte base-64 encoded string for data encryption.
 # Generate one with `docker run --rm cyberark/conjur data-key generate`.
-dataKey: ""
+# This is a required setting.
+# dataKey:
 
 # Annotations to apply to the Conjur deployment.
 deployment:
```

If the Helm linter CI passes with this change, can you please:
- Delete the added line in the CHANGELOG.md file (per @izgeri's earlier request)
- Squash the changes into one commit

Thank you!

**Pull request merged**